Google Container Engine cluster management module

erjohnso opened the pull request:
This module adds support to create / destroy / update Google Container Engine (GKE) Kubernetes clusters [[Product details here](https://cloud.google.com/container-engine/)].

Module DOCUMENTATION and EXAMPLES provide details for how to use the module.

This module can be combined with the (pending) module in #1229 to allow users to write a single playbook to spin up a Kubernetes cluster (on GKE) and deploy / manage their app with native Kubernetes primitives.

/cc @jimi-c, @gregdek, @bcoca

roberthbailey commented on clustering/gke.py:
```diff
@@ -0,0 +1,364 @@
+    description:
+      - The password for accessing the Kubernetes cluster
+    required: true
+    default: "correct_horse_battery_staple"
```
It'd be better not to have a default, well-known password here and instead force people to specify one (or alternatively, generate one randomly).

erjohnso replied:
Thanks @roberthbailey - I'll set it to `required` but not provide a default. That will force users to set their own (and they can generate one randomly if desired).

roberthbailey replied:
That would be much better. Thanks!

gregdek commented:
@erjohnso After some review, it looks as though this PR needs some revisions. 

Please make the suggested revisions, and when you're done, please add a comment with the text "ready_for_review" and we will continue with the review process.  Thanks!

robynbergeron commented:
@erjohnso A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes.

erjohnso commented:
Squashed commits, rebased to current devel, and made the following updates:
- Addressed feedback on passwords
- Bumped `version_added` to 2.1
- Added `RETURN` documentation
- Guard against missing required python module dep

ready_for_review

robynbergeron commented:
Thanks @erjohnso for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

roberthbailey commented:
LGTM.

roberthbailey commented:
shipit

detiber commented:
@erjohnso The separate 'update' state seems a bit counter-intuitive to me, I would expect that using 'present' and changing some of the parameters would be sufficient. 

Not being familiar with GKE though, I can't gell if desiredNodeVersion and desiredMonitoringService need special handling compared to the other values or if using state='present' will update the other values provided and a second invocation of the gke module would be needed to rev the NodeVersion and/or Monitoring Service for changes to take effect.

erjohnso commented:
Thanks for taking a look @detiber, very much appreciated!

While the module could be re-worked to only use 'present' as you suggest, I felt that a different action was safer for the user and more aligned with GKE product features. An 'upgrade' is an [invasive action](https://cloud.google.com/container-engine/docs/clusters/upgrade), so I think making the user specify an explicity action is preferable to being surprised.

bcoca commented:
@erjohnso yes, but when using ansible the 'present' = "create or update" pattern is expected.

erjohnso commented:
Ok, sounds like this should be put back in 'needs_review' then.

gregdek commented:
@erjohnso yep, back in needs_revision now. Thanks @bcoca and @detiber for the reviews.

erjohnso commented:
@detiber, @bcoca - sorry for sitting on this so long.

After thinking about this more and digging into GKE docs, I'm going to say that using `state=present` is _not_ the right fit for a cluster update.

[Creating a GKE cluster](https://cloud.google.com/container-engine/reference/rest/v1/projects.zones.clusters/create) requires a [`Cluster`](https://cloud.google.com/container-engine/reference/rest/v1/projects.zones.clusters#Cluster) resource. However, [Updating an existing GKE cluster](https://cloud.google.com/container-engine/reference/rest/v1/projects.zones.clusters/update) requires a [`ClusterUpdate`](https://cloud.google.com/container-engine/reference/rest/v1/projects.zones.clusters/update#ClusterUpdate) resource. Only one nested API resource, the [`AddonsConfig`](https://cloud.google.com/container-engine/reference/rest/v1/projects.zones.clusters#AddonsConfig) has overlap between the API's create and update operations.

So, from an Ansible user's point of view, I would need to specify a different set of module attributes depending on my use-case of create or update. I think, given both the possible destructive nature of an "update" (mentioned previously) combined with almost no re-use of module attributes, that it's fair to go out-of-band with this module and allow a `state=upgrade` choice.

However, if the community feels this is still not the right fit, then I suggest we remove the 'upgrade' functionality from the pull request. Instead, we provide documentation for Ansible users to use the [gcloud command-line](https://cloud.google.com/container-engine/docs/clusters/upgrade#the_upgrade_command) for doing cluster upgrades versus using Ansible.

wdyt?

detiber commented:
@erjohnso considering the disconnect between creating an initial cluster and update, would it make sense to have them be separate modules?

I'm thinking about the use case of ongoing management of an environment with Ansible, and having the present/upgrade difference seems more suited towards an ad-hoc use case. Considering the scenario, I could see having something broken down like this:
- gke_create
  - state:
    - present - create, do nothing if exists
    - absent - remove cluster
- gke_manage_cluster
  - state:
    - present - update existing cluster if parameters differ from existing cluster config
    - absent - remove cluster

The alternative is that the user would have to add the logic to create/update in some way on their own, probably by doing the following in a role to manage a gke cluster:

```
- name: Create a the GKE cluster
  gke:
    project_id: "{{ project_id }}"
    name: gke-test
    ...
    state: present
  register gke_create

  gke:
    project_id: "{{ project_id }}"
    name: gke-test
    ...
    state: upgrade
  when: not gke_create | changed
```

erjohnso commented:
@detiber - thanks for the quick response and feedback. I wish I could be as timely! :)

I was expecting users to have two separate playbooks depending on their need to create a cluster or do an upgrade. And in that scenario, I don't see that it makes a difference if they're using one module or two.

Honestly, I don't understand the aversion to using `state=upgrade`. It captures the intent of what the user wants to do in a clear way. Similarly, `state=present` captures the intent when a user wants to create a cluster.  But `state=present` when upgrading does not seem clear to me. Are we implying that an "upgrade is present"?

Also, if I look through existing modules, there are plenty of other cases where Ansible seems OK to go outside the normal `present` and `absent` choices.

```
$ grep -r state ansible-modules-* | grep -oP 'choices=\[.*\]' | sed -e 's|choices=\[||' -e 's|,| |g' -e 's|\]||' -e 's|,||g' -e "s|'||g" -e 's|"||g' | tr " " "\n" | grep -v ^$ | sort | uniq

absent
active
all
attached
build
build-dep
clean
configured
deleted
deployed
destroyed
detached
directory
disabled
drained
dump
enabled
enforcing
expunged
extracted
failed
file
finalize
finished
followed
hard
has_instances
import
inactive
installed
keys
killed
latest
link
list
locked
monitored
mounted
once
ongoing
opts_absent
opts_present
paused
permissive
present
query
read
rebalanced
reloaded
removed
reset
restart
restarted
restored
revert
running
shutdown
started
stopped
suspended
touch
undefined
unfollowed
unlocked
unmonitored
unmounted
updated
```

bcoca commented:
that is not something we encourage and are currently trying to reign in

erjohnso commented:
@bcoca - Gotcha. Trying to get more uniformity sounds like a great goal. Are there other standard choices you're considering beyond just create/destroy?  If so, is there something like replace or update (to match basic CRUD) that we can consider?

erjohnso commented:
@detiber, @bcoca - I ripped out `state=update`. This module only allows create/delete of a GKE cluster now.

Everyone's been great about reviewing already, but iiuc, @gregdek can flip this back to "community_review"?

gregdek commented:
Per new module review meeting on 6 April, this is shipit. Thanks @erjohnso for your patience here.

erjohnso commented:
Great, thanks @gregdek!

jimi-c commented on clustering/gke.py:
```diff
@@ -0,0 +1,366 @@
+            ),
+            node_config_metadata=dict(required=False),
+            username=dict(required=False, default="admin"),
+            password=dict(required=True),
```
The password param should be marked as `no_log=True` to prevent leaking it during logging.

jimi-c commented on clustering/gke.py:
```diff
@@ -0,0 +1,366 @@
+    if state == "present":
+        body = _make_cluster_body(module)
+        try:
+            resp = cluster_func().create(projectId=project, zone=zone,
```
Is this idempotent? Are you just relying on the service call making sure it doesn't recreate anything necessary, and will this update the cluster to match the required config if it does already exist?

roberthbailey replied:
If the cluster already exists the API will return a 409 (see the exception handling).

jimi-c replied:
Ok that's what I thought, that this relies on the API to handle idempotency.

erjohnso replied:
Is that sufficient, or should I query existing state (assuming there is a cluster), and if there are differences, destroy and re-create with new values?

roberthbailey replied:
It depends on what your expected behavior is. If you pass in the same project/zone/cluster name do you intend for that to always be the same cluster? Or do you need to check the other parameters like number of nodes to verify that the cluster is what you are trying to create? The API only checks the project/zone/name for conflicts, but your client may want/need to be more intelligent.

jimi-c replied:
I would expect it to update the cluster if it existed but had different params. It would be confusing to users to specify new settings and have an unchanged result be returned but to have a cluster which didn't match their expected settings.

gregdek commented:
@erjohnso A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@erjohnso Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@erjohnso A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@erjohnso Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@erjohnso A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gundalow commented:
Hi,

Looks like this branch needs updating to fix the Travis Issues, please see this guide on how to do that cleanly.

http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

needs_revision

gregdek commented:
Kicking this to bring to attention to new maintainer

gregdek commented:
Thanks @erjohnso for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@erjohnso A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@erjohnso Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@erjohnso A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@erjohnso Another friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open. If you have addressed the issues and believe it's ready for review, please comment with the text "ready_for_review". If we don't hear from you within another 14 days, we will close this pull request.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
So, looks like this is in never-never-land and has been for a while -- and I'm pretty sure that @erjohnso is no longer actively working on these.

Thus, I'm closing, and pinging @supertom as the heir apparent to @erjohnso -- maybe the best thing to do is fork and resubmit if needed.

erjohnso commented:
Thanks @gregdek! @supertom and I were discussing doing just that earlier this week!  He can definitely incorporate all the feedback on this PR and hopefully land a much better module soon.
